uberenv.py

```diff
@@ -1035,6 +1052,39 @@ def use_spack_upstream(self):
                 upstreams_cfg_file.write("  {0}:\n".format(upstream_name))
                 upstreams_cfg_file.write("    install_tree: {0}\n".format(upstream_path))
 
+    def setup_clingo(self):
+        """
+        Attempts to install the clingo answer set programming library
+        if it is not already available as a Python module
+        """
+        try:
+            import clingo
+        except ModuleNotFoundError:
+            import pip
+            pip_ver = pip.__version__
+            # Requirement comes from https://github.com/pypa/manylinux
+            # JBE: I think the string comparison is somewhat correct here, if not we'll
+            # need to install setuptools for 'packaging.version'
+            if pip_ver < "19.3":
+                print("[ERROR: pip version {0} is too old to install clingo".format(pip_ver))
+                print("  pip 19.3 is required for PEP 599 support")
+                print("]")
+                sys.exit(1)
+            py_interp = sys.executable
+            clingo_pkg = "clingo"
+            uninstall_cmd = "{0} -m pip uninstall -y {1}".format(py_interp, clingo_pkg)
+            # Uninstall it first in case the available version failed due to differing arch
+            # pip will still return 0 in the case of a "trivial" uninstall
+            res = sexe(uninstall_cmd, echo=True)
+            if res != 0:
+                print("[ERROR: clingo uninstall failed with returncode {0}]".format(res))
+                sys.exit(1)
+            install_cmd = "{0} -m pip install --user {1}".format(py_interp, clingo_pkg)
```

Comment:
since this will be a user install, this breaks the mold of uberenv a bit. But not sure there is an easy alternative.  If the system install already has clingo this path is shortcuted so thats a good option for folks that want to avoid user install.

Reply:
I tend to agree with @cyrush here.  

Directly calling `pip` on people's python seems different than the typical `uberenv` workflow. It might be preferable to error out here, with a message that tells people how to install `clingo` (and possibly provide some additional documentation about how to bootstrap their environment with `clingo`)

Reply:
@joshessman-llnl -- I was thinking about this a bit more and wanted to clarify my position/review comments.

This is a tricky problem (i.e. requiring a dependency before you can run the dependency-getting tool). 

While I don't like the tool pip-installing things into the user's python, I'm ok with the proposed solution as a stop-gap until we can find something better. Especially since (a) this is an opt-in solution -- i.e. the user has to specifically ask for `clingo`, and (b) there are issues getting newer versions of spack to concretize without `clingo` in some workflows.


In terms of something better, would it be possible for uberenv to install `clingo` into a `virtualenv`, which it could then use as its python for spack?  My main concern is that installing into a user's system python can have unexpected/undesired side effects.

**tl;dr;** -- I'm ok with the current solution, and will welcome a better solution in the future